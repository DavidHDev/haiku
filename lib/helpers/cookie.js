export const parseToDataType = (value, isItRetry = false) => {
  try {
    return value === 'undefined' ? undefined : JSON.parse(value ?? '');
  } catch (e) {
    if (!isItRetry) return parseToDataType(`"${value?.replaceAll?.('"', '')}"`, true);
    return undefined;
  }
};

export const parseToCookieType = (value) => {
  if(typeof value === 'string') return value;

  return JSON.stringify(value);
}

export const getCookie = (name) => {
  const value = `; ${document.cookie}`;

  const [, cookie] = value.split(`; ${name}=`);

  return cookie ? parseToDataType(cookie.split(';')[0]) : cookie;
};

export const getCookies = (cookies = []) => {
  if (cookies.length) return cookies.reduce((result, cookie) => ({ ...result, [cookie]: getCookie(cookie) }), {});

  return Object.fromEntries(document.cookie.split('; ').map((c) => {
    const [key, value] = c.split('=');
    return [key, parseToDataType(value)]
  }));
};

export const setCookie = (name, value, expireDays) => {
  const date = new Date();

  date.setTime(date.getTime() + expireDays * 24 * 60 * 60 * 1000);
  const expires = `expires=${date.toUTCString()}`;

  document.cookie = `${name}=${parseToCookieType(value)}; ${expires}; path=/;`;
};

export const deleteCookie = (name) => {
  document.cookie = `${name}=; path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;`;
};
