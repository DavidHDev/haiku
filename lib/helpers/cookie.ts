export const parseToDataType = <T>(
  value: string | undefined,
  isItRetry = false,
): T | undefined => {
  try {
    return value === 'undefined' || value === undefined
      ? undefined
      : (JSON.parse(value) as T);
  } catch (e) {
    if (!isItRetry) {
      return parseToDataType<T>(`"${value?.replaceAll?.('"', '')}"`, true);
    }
    return undefined;
  }
};

export const parseToCookieType = <T>(value: T): string => {
  if (typeof value === 'string') return value;
  return JSON.stringify(value);
};

export const getCookie = <T>(name: string): T | undefined => {
  const value = `; ${document.cookie}`;
  const [, cookie] = value.split(`; ${name}=`);

  return cookie ? parseToDataType<T>(cookie.split(';')[0]) : undefined;
};

export const getCookies = <T>(
  cookies: string[] = [],
): Record<string, T | undefined> => {
  if (cookies.length) {
    return cookies.reduce(
      (result, cookie) => ({ ...result, [cookie]: getCookie<T>(cookie) }),
      {} as Record<string, T | undefined>,
    );
  }

  return Object.fromEntries(
    document.cookie.split('; ').map((c) => {
      const [key, value] = c.split('=');

      return [key, parseToDataType<T>(value)];
    }),
  ) as Record<string, T | undefined>;
};

export const setCookie = <T>(
  name: string,
  value: T,
  expireDays: number,
): void => {
  const date = new Date();
  const millisecondsInADay = 24 * 60 * 60 * 1000;
  const expireDate = date.getTime() + expireDays * millisecondsInADay;

  date.setTime(expireDate);

  const expires = `expires=${date.toUTCString()}`;

  document.cookie = `${name}=${parseToCookieType(value)}; ${expires}; path=/;`;
};

export const deleteCookie = (name: string): void => {
  document.cookie = `${name}=; path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;`;
};
