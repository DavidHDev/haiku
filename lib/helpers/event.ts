export const on = <T extends EventTarget>(
  target: T,
  type: string,
  callback: EventListenerOrEventListenerObject | null,
  options?: AddEventListenerOptions | boolean
) => {
  if (target && target.addEventListener)
    target.addEventListener(type, callback, options);
};

export const off = <T extends EventTarget>(
  target: T,
  type: string,
  callback: EventListenerOrEventListenerObject | null,
  options?: AddEventListenerOptions | boolean
) => {
  if (target && target.removeEventListener)
    target.removeEventListener(type, callback, options);
};
