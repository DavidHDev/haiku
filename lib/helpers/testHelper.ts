import { ReactNode } from "react";
import { render as rtlRender } from "@testing-library/react";
import { expect } from "vitest";

export const takeSnapshot = (ui: ReactNode) => {
  const { asFragment } = rtlRender(ui);
  expect(asFragment()).toMatchSnapshot();
};
