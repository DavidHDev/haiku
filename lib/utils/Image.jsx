import React from "react";

export const Image = ({
  src,
  loading,
  alt,
  fallback,
  style,
  width,
  height,
  className,
}) => {
  const handleBrokenImage = (event) => (event.target.src = fallback);

  return (
    <img
      src={src}
      loading={loading}
      alt={alt}
      onError={handleBrokenImage}
      style={style}
      width={width}
      height={height}
      className={className}
    />
  );
};
