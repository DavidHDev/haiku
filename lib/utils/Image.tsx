import { ImgHTMLAttributes, SyntheticEvent } from 'react';

type Props = ImgHTMLAttributes<HTMLImageElement> & { fallback: string };

export const Image = ({ src, loading, alt, fallback, ...props }: Props) => {
  const handleBrokenImage = (event: SyntheticEvent<HTMLImageElement, Event>) =>
    (event.currentTarget.src = fallback);

  return (
    <img
      src={src}
      loading={loading}
      alt={alt}
      onError={handleBrokenImage}
      {...props}
    />
  );
};
