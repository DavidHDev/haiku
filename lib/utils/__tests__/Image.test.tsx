import { describe, expect, it } from 'vitest';
import { takeSnapshot } from 'lib/helpers/testHelper';
import { fireEvent, render, screen } from '@testing-library/react';
import { Image } from '../Image';

describe('Image', () => {
  const src = 'https://example.com/image.jpg';
  const fallback = 'https://example.com/fallback.jpg';
  const alt = 'Example Image';

  it('renders the Image component', () => {
    takeSnapshot(<Image alt={alt} src={src} fallback={fallback} />);
    const imgElement = screen.getByAltText(alt);
    expect(imgElement).toBeInTheDocument();
    expect(imgElement).toHaveAttribute('src', src);
  });

  it('renders with loading attribute', () => {
    const loading = 'lazy';
    render(<Image src={src} alt={alt} loading={loading} fallback={fallback} />);
    const imgElement = screen.getByAltText(alt);
    expect(imgElement).toHaveAttribute('loading', loading);
  });

  it('renders with additional props', () => {
    const cutomClass = 'cutom-class';
    render(
      <Image src={src} alt={alt} fallback={fallback} className={cutomClass} />,
    );
    const imgElement = screen.getByAltText(alt);
    expect(imgElement).toHaveClass(cutomClass);
  });

  it('handles broken image by setting fallback src', () => {
    render(<Image src={src} alt={alt} fallback={fallback} />);
    const imgElement = screen.queryByRole('img');
    fireEvent.error(imgElement!);
    expect(imgElement).toHaveAttribute('src', fallback);
  });

  it('should not change src if image loads correctly', () => {
    render(<Image src={src} alt={alt} fallback={fallback} />);
    const imgElement = screen.queryByRole('img');
    fireEvent.load(imgElement!);
    expect(imgElement).toHaveAttribute('src', src);
  });
});
