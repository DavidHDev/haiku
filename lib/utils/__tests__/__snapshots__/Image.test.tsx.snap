// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Image > renders the Image component 1`] = `
<DocumentFragment>
  <img
    alt="Example Image"
    src="https://example.com/image.jpg"
  />
</DocumentFragment>
`;
